src/passes/OptimizeInstructions.cpp

```diff
@@ -893,6 +907,32 @@ struct OptimizeInstructions
           }
         }
       }
+      {
+        // !!x  =>  x != 0
+        Expression* x;
+        if (matches(curr, unary(EqZ, unary(EqZ, any(&x))))) {
+          auto xType = x->type;
+
+          // !!bool(x)  ==>   bool(x)
+          // Note that this is only possible if the value is the same type,
+          // which is not the case for (i32.eqz (i64.eqz ..)).
+          if (xType == curr->type && Bits::getMaxBits(x, this) == 1) {
+            return replaceCurrent(x);
+          }
+
+          // If the input is not boolean already (and it does not change type
+          // from i64 to i32), then we must add a ! = 0 which adds one byte, but
+          // it improves speed as it swaps two CPU instructions with one. Even
+          // when moderately optimizing for size this is reasonable to do, as
+          // the code size increase is tiny compared to the potential benefits.
+          if (getPassOptions().shrinkLevel <= 1) {
```

Comment:
When optimizing for size, should we continue doing the opposite transformation?

Reply:
Perhaps yes, but for now I didn't add that. I can leave a TODO.